### PR TITLE
Excise SubpageId and use only PipelineIds

### LINKED ISSUE #11694: Remove SubpageId
This will help me with #10808.

### PR DESCRIPTION
**Pull request by aneeshusa**
SubpageId was originally introduced in 2013 to help iframes keep track of
their associated (children) pipelines. However, since each pipeline
already has a PipelineId, and those are unique, those are sufficient
to keep track of children.

---

- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [X] These changes fix #11694  (github issue number if applicable).

- [ ] There are tests for these changes OR
- [X] These changes do not require tests because refactoring

---

This change is [Reviewable](https://reviewable.io/reviews/servo/servo/11698)

### CODE REVIEW

**Comment by aneeshusa** [comment]
BTW, this isn't a rebase of #7792; I learned a lot about how Servo works by doing this myself :)

**Comment by ConnorGBrewster** [comment]
Would it be possible to get rid of the `(PipelineId, PipelineId)`s?

**Comment by aneeshusa** [comment]
Rebased (not sure why highfive didn't leave a comment).

@ConnorGBrewster in general, the second PipelineId is necessary to identify which pipeline the message is about, and the first PipelineId is for the parent, which is needed to find the relevant iframe via `get_browsing_context()` -> `activate_document()` -> `find_iframe()`.

It's possible that in some cases this may be able to be reduced to just the child PipelineId, but it's not obvious to me that that holds and I'd like to leave that for a follow-up.

**Comment by ConnorGBrewster** [comment]
@aneeshusa makes sense, I was thinking you could just do the child and do the lookup in the constellation. But yeah that seems like it would be out of scope for this PR.

File: components/script/script_thread.rs [hunk 1]
```diff
@@ -1969,21 +1963,22 @@ impl ScriptThread {
...
         {
             let nurl = &load_data.url;
             if let Some(fragment) = nurl.fragment() {
-                let document = match self.root_browsing_context().find(pipeline_id) {
+                let document = match self.root_browsing_context().find(parent_pipeline_id) {
                     Some(browsing_context) => browsing_context.active_document(),
-                    None => return warn!("Message sent to closed pipeline {}.", pipeline_id),
+                    None => return warn!("Message sent to closed pipeline {}.", parent_pipeline_id),
                 };
+                let document = document.r();
```
Comment (aneeshusa) [inline components/script/script_thread.rs]: Note: @asajeffrey removed this line in #11695, but this seemed inadvertent to me so I brought it back. Not sure about this change.

Comment (Ms2ger) [inline components/script/script_thread.rs]: Please remove it again.

**Comment by ConnorGBrewster** [comment]
This is looking awesome! I have a couple things for you to look at, but other than that, it looks good. I especially like the move to more consistent naming.

I would like to see if there is a way to avoid passing parent and child pipelines everywhere and I would like to have `FrameType` moved to the `Pipeline` itself, but I think both those can be done in separate PRs.

After the fixes, it would probably be good to have someone else do a look over the changes.

---

Reviewed 7 of 17 files at r1, 1 of 13 files at r7, 4 of 16 files at r10, 11 of 12 files at r12, 1 of 1 files at r13.
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[components/constellation/constellation.rs, line 1363 \[r13\]](https://reviewable.io:443/reviews/servo/servo/11698#-KLDkNpm_O1ovdAiz592:-KLDkNpm_O1ovdAiz593:-1609851760) ([raw file](https://github.com/servo/servo/blob/19cf404d1105f17fd2284e12be9f6494797be914/components/constellation/constellation.rs#L1363)):*
> ```Rust
>         // This makes things like contentDocument work correctly.
>         if let Some((parent_pipeline_id, _)) = pipeline_info {
>             match self.pipelines.get(&next_pipeline_id) {
> ```

This match can go away.

---

*[components/script/script_thread.rs, line 1197 \[r13\]](https://reviewable.io:443/reviews/servo/servo/11698#-KLDsidMtKNLUUAaPx_F:-KLDsidMtKNLUUAaPx_G:-1150341064) ([raw file](https://github.com/servo/servo/blob/19cf404d1105f17fd2284e12be9f6494797be914/components/script/script_thread.rs#L1197)):*
> ```Rust
> 
>         // Kick off the fetch for the new resource.
>         let new_load = InProgressLoad::new(new_pipeline_id, Some((parent_pipeline_id, frame_type)),
> ```

Should this be the child pipeline id instead of the parent?

---

*[components/script/script_thread.rs, line 1427 \[r13\]](https://reviewable.io:443/reviews/servo/servo/11698#-KLDswiyN6yHMmQs4an7:-KLDswiyN6yHMmQs4an8:-345657493) ([raw file](https://github.com/servo/servo/blob/19cf404d1105f17fd2284e12be9f6494797be914/components/script/script_thread.rs#L1427)):*
> ```Rust
>     fn handle_page_headers_available(&self, id: &PipelineId,
>                                      metadata: Option<Metadata>) -> Option<ParserRoot> {
>         let idx = self.incomplete_loads.borrow().iter().position(|load| { load.pipeline_id == *id });
> ```

See above about child pipeline instead of parent.

---

*[components/script/dom/htmliframeelement.rs, line 226 \[r13\]](https://reviewable.io:443/reviews/servo/servo/11698#-KLDum0gBXZ_jaZbHiOQ:-KLDum0gBXZ_jaZbHiOR:-1512445507) ([raw file](https://github.com/servo/servo/blob/8dcb12a019885d8333f35f43e43d3f2e0c56573a/components/script/dom/htmliframeelement.rs#L226)):*
> ```Rust
>     }
> 
>     pub fn pipeline(&self) -> Option<PipelineId> {
> ```

That's weird that we had both `pipeline()` and `pipeline_id()`...

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/11698#-:-KLDvTqsT9d9ssDf6qwe:-1371215943)*

**Comment by aneeshusa** [comment]
Rebased.

I'm not sure what you mean about removing the `match` - what should I replace it with?

The two checks in the original `handle_page_headers_available` were redundant - the first was checking that the pipeline id matched, and the second that the subpage id matched, so I've kept just the first. The id in `handle_page_headers_available` already is the `PipelineId` of the child. The other place you mentioned using the child id is just passing the `parent_info` along, which always has `parent_pipeline_id` as the first member.

**Comment by ConnorGBrewster** [comment]
That match was used to find `new_subpage_id` which is no longer needed for sending the `UpdateSubpageId` message, so the match can just be removed. The only purpose I could see is for checking for the existence of the pipeline/parent, but I don't think we need to do that there.

**Comment by aneeshusa** [comment]
OK, removed it.

**Comment by ConnorGBrewster** [comment]
Ignore my comments about the child pipeline, after looking it over again, it looks fine. After a rebase I think these changes look good, but it would be a good idea to have @asajeffrey sign off on this.

r? @asajeffrey 

---

Reviewed 4 of 16 files at r14, 11 of 12 files at r16, 1 of 1 files at r17.
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/11698#-:-KLrTK3jFxeVOgFdnVXg:1902757492)*

**Comment by asajeffrey** [comment]
Catching up on reviewing... Mostly this looks really good, and I'm glad to see the back of subpage ids. There are a bunch of places where I think we should use `FrameId`s though, in general anywhere where we used to have `(PipelineId, SubpageId)` we should now have `FrameId`. (Note, this assumes frame ids are globally unique.)

---

Reviewed 3 of 17 files at r1, 5 of 20 files at r18, 4 of 16 files at r21, 12 of 12 files at r23.
Review status: all files reviewed at latest revision, 23 unresolved discussions, some commit checks failed.

---

*[components/constellation/constellation.rs, line 30 \[r23\]](https://reviewable.io:443/reviews/servo/servo/11698#-KMZJdOlO-yOLAvIcv_U:-KMZJdOmMkQVzZtT7f-A:-1502557946) ([raw file](https://github.com/servo/servo/blob/abdf2f28a0bd25a76a31b4fd2410c7a87448ed54/components/constellation/constellation.rs#L30)):*
> ```Rust
> use msg::constellation_msg::{Key, KeyModifiers, KeyState, LoadData};
> use msg::constellation_msg::{PipelineNamespace, PipelineNamespaceId, NavigationDirection};
> use msg::constellation_msg::{SubpageId, WindowSizeType};
> ```

Yay, no more subpage ids!

---

*[components/constellation/constellation.rs, line 1277 \[r23\]](https://reviewable.io:443/reviews/servo/servo/11698#-KMZK0qNlauxvDvMIWQS:-KMZK0qNlauxvDvMIWQT:-1750647076) ([raw file](https://github.com/servo/servo/blob/a79a1dffcbef22e7699d7eadc493271a2804024f/components/constellation/constellation.rs#L1277)):*
> ```Rust
> 
>     fn handle_navigate_msg(&mut self,
>                            pipeline_info: Option<(PipelineId, PipelineId)>,
> ```

Should this not be a `FrameId`?

---

*[components/constellation/constellation.rs, line 1437 \[r23\]](https://reviewable.io:443/reviews/servo/servo/11698#-KMZKGCmja7SJ_2sDwPn:-KMZKGCmja7SJ_2sDwPo:-560546932) ([raw file](https://github.com/servo/servo/blob/a79a1dffcbef22e7699d7eadc493271a2804024f/components/constellation/constellation.rs#L1437)):*
> ```Rust
>     fn handle_mozbrowser_event_msg(&mut self,
>                                    parent_pipeline_id: PipelineId,
>                                    pipeline_id: PipelineId,
> ```

`FrameId`?

---

*[components/constellation/pipeline.rs, line 377 \[r23\]](https://reviewable.io:443/reviews/servo/servo/11698#-KMZKgazZGfbs2VCZkR_:-KMZKgazZGfbs2VCZkRa:-560546932) ([raw file](https://github.com/servo/servo/blob/a79a1dffcbef22e7699d7eadc493271a2804024f/components/constellation/pipeline.rs#L377)):*
> ```Rust
> 
>     pub fn trigger_mozbrowser_event(&self,
>                                      child_id: PipelineId,
> ```

`FrameId`?

---

*[components/script/dom/browsingcontext.rs, line 161 \[r23\]](https://reviewable.io:443/reviews/servo/servo/11698#-KMZLCcBLikMafqEnvzt:-KMZLCcBLikMafqEnvzu:-560546932) ([raw file](https://github.com/servo/servo/blob/a79a1dffcbef22e7699d7eadc493271a2804024f/components/script/dom/browsingcontext.rs#L161)):*
> ```Rust
>     }
> 
>     pub fn find_child_by_id(&self, pipeline_id: PipelineId) -> Option<Root<Window>> {
> ```

`FrameId`?

---

*[components/script/dom/document.rs, line 1499 \[r23\]](https://reviewable.io:443/reviews/servo/servo/11698#-KMZLOO8NoLrBadwqoeO:-KMZLOO9izqyI58Ti2FC:-560546932) ([raw file](https://github.com/servo/servo/blob/a79a1dffcbef22e7699d7eadc493271a2804024f/components/script/dom/document.rs#L1499)):*
> ```Rust
> 
>     /// Find an iframe element in the document.
>     pub fn find_iframe(&self, pipeline: PipelineId) -> Option<Root<HTMLIFrameElement>> {
> ```

`FrameId`?

---

*[components/script/dom/htmliframeelement.rs, line 70 \[r23\]](https://reviewable.io:443/reviews/servo/servo/11698#-KMZM3WyPeURuItWXcbP:-KMZM3WyPeURuItWXcbQ:409008752) ([raw file](https://github.com/servo/servo/blob/a79a1dffcbef22e7699d7eadc493271a2804024f/components/script/dom/htmliframeelement.rs#L70)):*
> ```Rust
> pub struct HTMLIFrameElement {
>     htmlelement: HTMLElement,
>     pipeline_id: Cell<Option<PipelineId>>,
> ```

Making sure I'm following this... The `pipeline_id` is mutable, and is the id of the current pipeline for the iframe. Do we not also need an immutable `frame_id`?

---

*[components/script/dom/htmliframeelement.rs, line 96 \[r23\]](https://reviewable.io:443/reviews/servo/servo/11698#-KMZMZFZidYNpA5aa7AD:-KMZMZFZidYNpA5aa7AE:62211560) ([raw file](https://github.com/servo/servo/blob/a79a1dffcbef22e7699d7eadc493271a2804024f/components/script/dom/htmliframeelement.rs#L96)):*
> ```Rust
>     }
> 
>     pub fn generate_new_pipeline_id(&self) -> (Option<PipelineId>, PipelineId) {
> ```

Is this function only used in `navigate_or_reload_child_browsing_context`? If so, can we get rid of it and inline it?

---

*[components/script/dom/window.rs, line 199 \[r23\]](https://reviewable.io:443/reviews/servo/servo/11698#-KMZNIwcbmOfdtwvJgxO:-KMZNIwcbmOfdtwvJgxP:1305257226) ([raw file](https://github.com/servo/servo/blob/a79a1dffcbef22e7699d7eadc493271a2804024f/components/script/dom/window.rs#L199)):*
> ```Rust
> 
>     /// Pipeline id associated with this page.
>     id: PipelineId,
> ```

Do we need to store the frame id as well?

---

*[components/script_traits/lib.rs, line 130 \[r23\]](https://reviewable.io:443/reviews/servo/servo/11698#-KMZNbFwlmTHmE2A2D8I:-KMZNbFwlmTHmE2A2D8J:1020245254) ([raw file](https://github.com/servo/servo/blob/a79a1dffcbef22e7699d7eadc493271a2804024f/components/script_traits/lib.rs#L130)):*
> ```Rust
>     pub parent_pipeline_id: PipelineId,
>     /// Id of the newly-created pipeline.
>     pub new_pipeline_id: PipelineId,
> ```

Do we need the frame id too?

---

*[components/script_traits/lib.rs, line 177 \[r23\]](https://reviewable.io:443/reviews/servo/servo/11698#-KMZO6JbxC7UoU97AlLB:-KMZO6JbxC7UoU97AlLC:2098316447) ([raw file](https://github.com/servo/servo/blob/a79a1dffcbef22e7699d7eadc493271a2804024f/components/script_traits/lib.rs#L177)):*
> ```Rust
>     /// Notifies script thread that a url should be loaded in this iframe.
>     /// First PipelineId is for the parent, second PipelineId is for the actual pipeline.
>     Navigate(PipelineId, PipelineId, LoadData),
> ```

Shouldn't this be `Navigate(FrameId, LoadData),`?

---

*[components/script_traits/lib.rs, line 180 \[r23\]](https://reviewable.io:443/reviews/servo/servo/11698#-KMZNpfkUYJIbZaV4dt5:-KMZNpfkUYJIbZaV4dt6:503895958) ([raw file](https://github.com/servo/servo/blob/a79a1dffcbef22e7699d7eadc493271a2804024f/components/script_traits/lib.rs#L180)):*
> ```Rust
>     /// Requests the script thread forward a mozbrowser event to an iframe it owns
>     /// First PipelineId is for the parent, second PipelineId is for the actual pipeline.
>     MozBrowserEvent(PipelineId, PipelineId, MozBrowserEvent),
> ```

Shouldn't this be `MozBrowserEvent(FrameId, MozBrowserEvent),`?

---

*[components/script_traits/lib.rs, line 184 \[r23\]](https://reviewable.io:443/reviews/servo/servo/11698#-KMZOL-712Iy7NK27UHM:-KMZOL-712Iy7NK27UHN:-1843920607) ([raw file](https://github.com/servo/servo/blob/a79a1dffcbef22e7699d7eadc493271a2804024f/components/script_traits/lib.rs#L184)):*
> ```Rust
>     /// First PipelineId is for the parent, second is the old PipelineId for the frame,
>     /// third is the new PipelineId for the frame.
>     UpdatePipelineId(PipelineId, PipelineId, PipelineId),
> ```

Shouldn't this be `UpdatePipelineId(FrameId, PipelineId),`?

---

*[components/script_traits/lib.rs, line 186 \[r23\]](https://reviewable.io:443/reviews/servo/servo/11698#-KMZOWnWC3Uad_gxuc6n:-KMZOWnWC3Uad_gxuc6o:794237684) ([raw file](https://github.com/servo/servo/blob/a79a1dffcbef22e7699d7eadc493271a2804024f/components/script_traits/lib.rs#L186)):*
> ```Rust
>     UpdatePipelineId(PipelineId, PipelineId, PipelineId),
>     /// Set an iframe to be focused. Used when an element in an iframe gains focus.
>     FocusIFrame(PipelineId, PipelineId),
> ```

Shouldn't this be `FocusIFrame(FrameId)`?

---

*[components/script_traits/lib.rs, line 202 \[r23\]](https://reviewable.io:443/reviews/servo/servo/11698#-KMZOgxgzig5lhYmtLyU:-KMZOgxgzig5lhYmtLyV:2019484334) ([raw file](https://github.com/servo/servo/blob/a79a1dffcbef22e7699d7eadc493271a2804024f/components/script_traits/lib.rs#L202)):*
> ```Rust
>     },
>     /// Notifies a parent frame that one of its child frames is now active.
>     FramedContentChanged(PipelineId, PipelineId),
> ```

Shouldn't this be `FramedContentChanged(FrameId)`?

---

*[components/script_traits/lib.rs, line 594 \[r23\]](https://reviewable.io:443/reviews/servo/servo/11698#-KMZOpRLv3WY1Ar-KEf8:-KMZOpRLv3WY1Ar-KEf9:-861254037) ([raw file](https://github.com/servo/servo/blob/a79a1dffcbef22e7699d7eadc493271a2804024f/components/script_traits/lib.rs#L594)):*
> ```Rust
>     /// Request to navigate a frame.
>     /// First PipelineId is for the window, second PipelineId is for the iframe.
>     Navigate(Option<(PipelineId, PipelineId)>, NavigationDirection),
> ```

Shouldn't this be `Navigate(Option<FrameId>, NavigationDirection),`?

---

*[components/script_traits/script_msg.rs, line 76 \[r23\]](https://reviewable.io:443/reviews/servo/servo/11698#-KMZP5__KPd58E9D0JVK:-KMZP5__KPd58E9D0JVL:1700997001) ([raw file](https://github.com/servo/servo/blob/a79a1dffcbef22e7699d7eadc493271a2804024f/components/script_traits/script_msg.rs#L76)):*
> ```Rust
>     /// Dispatch a mozbrowser event to a given iframe. Only available in experimental mode.
>     /// First PipelineId is for the parent, second PipelineId is for the actual pipeline.
>     MozBrowserEvent(PipelineId, PipelineId, MozBrowserEvent),
> ```

Ditto.

---

*[components/script_traits/script_msg.rs, line 79 \[r23\]](https://reviewable.io:443/reviews/servo/servo/11698#-KMZP7fqwSxZr5R7HMaT:-KMZP7fqwSxZr5R7HMaU:1700997001) ([raw file](https://github.com/servo/servo/blob/a79a1dffcbef22e7699d7eadc493271a2804024f/components/script_traits/script_msg.rs#L79)):*
> ```Rust
>     /// HTMLIFrameElement Forward or Back navigation.
>     /// First PipelineId is for the window, second PipelineId is for the iframe.
>     Navigate(Option<(PipelineId, PipelineId)>, NavigationDirection),
> ```

Ditto.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/11698#-:-KMZPRPQPW_4KbGV_m7Z:1731490788)*

**Comment by aneeshusa** [comment]
It looks like `SubpageId`s are created well before the corresponding `FrameId` ever is.

Also, it does look like `FrameId`s are globally unique, but they appear to be systematically generated (auto-incrementing counter). If I change all of these events/messages to use `FrameId`s, then wouldn't it be possible to guess/spoof a `FrameId` and send a message about a different frame? Would `FrameId`s also need to be UUIDs for security purposes?

I'd like to stick with `PipelineIds` for now and think about replacing them with `FrameId`s later, since it looks like it'll be non-trivial.

**Comment by aneeshusa** [comment]
Follow up to last comment: it looks like `FrameId`s are mostly used for history tracking atm, and each Pipeline stores its `FrameId` (see #11809). I think it makes sense to keep using `PipelineId`s as general purpose identifiers here and keep `FrameId`s with a smaller scope of responsibility, so no need to add `FrameId`s to those two structs you mentioned above.

**Comment by asajeffrey** [comment]
Pipeline ids and frame ids identify different things. A pipeline is associated with a document, a frame id is associated with a browsing context (which contains a sequence of documents). In DOM terms, a pipeline is a Window, and a frame is an IFrame. The current `(pipeline_id, subpage_id)` lookups are identifying a frame, not a pipeline. For example:
```
        /// Dispatch a mozbrowser event to a given iframe. Only available in experimental mode.
        /// First PipelineId is for the parent, second PipelineId is for the actual pipeline.
        MozBrowserEvent(PipelineId, PipelineId, MozBrowserEvent),
```
should probably be:
```
        /// Dispatch a mozbrowser event to a given iframe. Only available in experimental mode.
        MozBrowserEvent(FrameId, MozBrowserEvent),
```
(either that, or the comment is inaccurate).

**Comment by asajeffrey** [comment]
Review status: all files reviewed at latest revision, 24 unresolved discussions, some commit checks failed.

---

*[components/constellation/constellation.rs, line 675 \[r23\]](https://reviewable.io:443/reviews/servo/servo/11698#-KN9Ip0SHh7pM79EgaXC:-KN9Ip0Tou4th3jr7heb:-1816019901) ([raw file](https://github.com/servo/servo/blob/a79a1dffcbef22e7699d7eadc493271a2804024f/components/constellation/constellation.rs#L675)):*
> ```Rust
>                 }
>             }
>             FromScriptMsg::MozBrowserEvent(parent_pipeline_id, pipeline_id, event) => {
> ```

Do we need the `parent_pipeline_id` here?

---

*[components/constellation/constellation.rs, line 1437 \[r23\]](https://reviewable.io:443/reviews/servo/servo/11698#-KMZKGCmja7SJ_2sDwPn:-KN9JaFynpuRC-HOZQQ-:5173665) ([raw file](https://github.com/servo/servo/blob/a79a1dffcbef22e7699d7eadc493271a2804024f/components/constellation/constellation.rs#L1437)):*
<details><summary><i>Previously, asajeffrey (Alan Jeffrey) wrote…</i></summary>
> `FrameId`?
</details>
OK

---

*[components/constellation/pipeline.rs, line 377 \[r23\]](https://reviewable.io:443/reviews/servo/servo/11698#-KMZKgazZGfbs2VCZkR_:-KN9KaA5ZbuiWb61mrZG:5173665) ([raw file](https://github.com/servo/servo/blob/a79a1dffcbef22e7699d7eadc493271a2804024f/components/constellation/pipeline.rs#L377)):*
<details><summary><i>Previously, asajeffrey (Alan Jeffrey) wrote…</i></summary>
> `FrameId`?
</details>
OK

---

*[components/script/dom/browsingcontext.rs, line 161 \[r23\]](https://reviewable.io:443/reviews/servo/servo/11698#-KMZLCcBLikMafqEnvzt:-KN9Kw9eqSs5_-jkVffD:5173665) ([raw file](https://github.com/servo/servo/blob/a79a1dffcbef22e7699d7eadc493271a2804024f/components/script/dom/browsingcontext.rs#L161)):*
<details><summary><i>Previously, asajeffrey (Alan Jeffrey) wrote…</i></summary>
> `FrameId`?
</details>
OK

---

*[components/script/dom/document.rs, line 1499 \[r23\]](https://reviewable.io:443/reviews/servo/servo/11698#-KMZLOO8NoLrBadwqoeO:-KN9L28QjhqOsT8N-9gT:5173665) ([raw file](https://github.com/servo/servo/blob/a79a1dffcbef22e7699d7eadc493271a2804024f/components/script/dom/document.rs#L1499)):*
<details><summary><i>Previously, asajeffrey (Alan Jeffrey) wrote…</i></summary>
> `FrameId`?
</details>
OK

---

*[components/script/dom/htmliframeelement.rs, line 70 \[r23\]](https://reviewable.io:443/reviews/servo/servo/11698#-KMZM3WyPeURuItWXcbP:-KN9LAjBM-iPF6YMikOj:-66361399) ([raw file](https://github.com/servo/servo/blob/a79a1dffcbef22e7699d7eadc493271a2804024f/components/script/dom/htmliframeelement.rs#L70)):*
<details><summary><i>Previously, asajeffrey (Alan Jeffrey) wrote…</i></summary>
> Making sure I'm following this... The `pipeline_id` is mutable, and is the id of the current pipeline for the iframe. Do we not also need an immutable `frame_id`?
</details>
Include the frame id in the HTMLIFrameElement object?

---

*[components/script/dom/window.rs, line 199 \[r23\]](https://reviewable.io:443/reviews/servo/servo/11698#-KMZNIwcbmOfdtwvJgxO:-KN9LYT987ViQMNBjYV5:5173665) ([raw file](https://github.com/servo/servo/blob/a79a1dffcbef22e7699d7eadc493271a2804024f/components/script/dom/window.rs#L199)):*
<details><summary><i>Previously, asajeffrey (Alan Jeffrey) wrote…</i></summary>
> Do we need to store the frame id as well?
</details>
OK

---

*[components/script_traits/lib.rs, line 177 \[r23\]](https://reviewable.io:443/reviews/servo/servo/11698#-KMZO6JbxC7UoU97AlLB:-KN9LhZXZAr_fiFi1E8A:-2028101843) ([raw file](https://github.com/servo/servo/blob/a79a1dffcbef22e7699d7eadc493271a2804024f/components/script_traits/lib.rs#L177)):*
<details><summary><i>Previously, asajeffrey (Alan Jeffrey) wrote…</i></summary>
> Shouldn't this be `Navigate(FrameId, LoadData),`?
</details>
It's not obvious why we need two pipeline ids, isn't the first one always the parent of the second?

---

*[components/script_traits/lib.rs, line 180 \[r23\]](https://reviewable.io:443/reviews/servo/servo/11698#-KMZNpfkUYJIbZaV4dt5:-KN9LrUKOvuaM4lzfary:1700997001) ([raw file](https://github.com/servo/servo/blob/a79a1dffcbef22e7699d7eadc493271a2804024f/components/script_traits/lib.rs#L180)):*
<details><summary><i>Previously, asajeffrey (Alan Jeffrey) wrote…</i></summary>
> Shouldn't this be `MozBrowserEvent(FrameId, MozBrowserEvent),`?
</details>
Ditto.

---

*[components/script_traits/lib.rs, line 184 \[r23\]](https://reviewable.io:443/reviews/servo/servo/11698#-KMZOL-712Iy7NK27UHM:-KN9LwOqvIWBo8NGVdIY:1700997001) ([raw file](https://github.com/servo/servo/blob/a79a1dffcbef22e7699d7eadc493271a2804024f/components/script_traits/lib.rs#L184)):*
<details><summary><i>Previously, asajeffrey (Alan Jeffrey) wrote…</i></summary>
> Shouldn't this be `UpdatePipelineId(FrameId, PipelineId),`?
</details>
Ditto.

---

*[components/script_traits/lib.rs, line 186 \[r23\]](https://reviewable.io:443/reviews/servo/servo/11698#-KMZOWnWC3Uad_gxuc6n:-KN9M-D9npNJN3zN18TF:1700997001) ([raw file](https://github.com/servo/servo/blob/a79a1dffcbef22e7699d7eadc493271a2804024f/components/script_traits/lib.rs#L186)):*
<details><summary><i>Previously, asajeffrey (Alan Jeffrey) wrote…</i></summary>
> Shouldn't this be `FocusIFrame(FrameId)`?
</details>
Ditto.

---

*[components/script_traits/lib.rs, line 202 \[r23\]](https://reviewable.io:443/reviews/servo/servo/11698#-KMZOgxgzig5lhYmtLyU:-KN9MN_fNNoT4w01TEfw:1700997001) ([raw file](https://github.com/servo/servo/blob/a79a1dffcbef22e7699d7eadc493271a2804024f/components/script_traits/lib.rs#L202)):*
<details><summary><i>Previously, asajeffrey (Alan Jeffrey) wrote…</i></summary>
> Shouldn't this be `FramedContentChanged(FrameId)`?
</details>
Ditto.

---

*[components/script_traits/lib.rs, line 594 \[r23\]](https://reviewable.io:443/reviews/servo/servo/11698#-KMZOpRLv3WY1Ar-KEf8:-KN9MRr_4xguSJRbDuss:1700997001) ([raw file](https://github.com/servo/servo/blob/a79a1dffcbef22e7699d7eadc493271a2804024f/components/script_traits/lib.rs#L594)):*
<details><summary><i>Previously, asajeffrey (Alan Jeffrey) wrote…</i></summary>
> Shouldn't this be `Navigate(Option<FrameId>, NavigationDirection),`?
</details>
Ditto.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/11698)*

**Comment by asajeffrey** [comment]
There are still some places where we're passing around two pipeline ids, and I don't really understand why: is the first not always the parent of the second?

**Comment by aneeshusa** [comment]
@asajeffrey That's right, I just didn't think it would trivial to shave it down to just the child id and wanted to get this landed and save it for a follow up. That being said, I've taken a stab at it in the last commit but I'm not 100% sure it's correct. In particular, it looks like all the messages I changed were being sent to the `script_chan` of the parent pipline, making the `.find(parent_pipeline_id)` calls in `script_thread.rs` no-ops (they just return the same `BrowsingContext`), and I removed them.

Re: adding `frame_id` to `HTMLIFrameElement`, AFAICT `FrameIds` are mainly used in the `constellation` crate, not `script` at all yet, and I don't see any need to add it (what would it be used for?). Also, `FrameId`s seem to be created much later than the corresponding `SubpageId`/`PipelineId`, so I'd have to add new plumbing to inject/update the `frame_id` variable, and I'd rather save that for a follow-up.

File: components/constellation/constellation.rs [hunk 2]
```diff
@@ -1521,20 +1508,15 @@ impl<Message, LTF, STF> Constellation<Message, LTF, STF>
...
-            let new_subpage_id = match self.pipelines.get(&next_pipeline_id) {
-                None => return warn!("Pipeline {:?} navigated to after closure.", next_pipeline_id),
-                Some(pipeline) => match pipeline.parent_info {
-                    None => return warn!("Pipeline {:?} has no parent info.", next_pipeline_id),
-                    Some((_, new_subpage_id, _)) => new_subpage_id,
-                },
+        if let Some(pipeline_id) = pipeline_info {
+            let parent_pipeline_id = match self.pipelines.get(&pipeline_id) {
+                None => return warn!("Pipeline {:?} navigated after closure.", pipeline_id),
+                Some(pipeline) => pipeline.parent_info.unwrap().0,
```
Comment (KiChjang) [inline components/constellation/constellation.rs]: Travis is complaining about this unwrap.

**Comment by asajeffrey** [comment]
@aneeshusa once you remove the unwrap, rebase and squash, you can r=me.

**Comment by KiChjang** [comment]
```
Ran 6831 tests finished in 876.0 seconds.
  • 6817 ran as expected. 1998 tests skipped.
  • 14 tests timed out unexpectedly

Tests with unexpected results:
  ▶ TIMEOUT [expected OK] /html/browsers/history/joint-session-history/joint-session-history-only-fully-active.html
  │ 
  │ Xlib:  extension "XFree86-VidModeExtension" missing on display ":0".
  └ ERROR:script::dom::bindings::error: Error at http://web-platform.test:8000/html/browsers/history/joint-session-history/joint-session-history-grandchild2.html:2:3 window.frameElement is null

  ▶ TIMEOUT [expected OK] /html/browsers/history/the-history-interface/001.html
  │ 
  │ Xlib:  extension "XFree86-VidModeExtension" missing on display ":0".
  └ ERROR:script::dom::bindings::error: Error at http://web-platform.test:8000/html/browsers/history/the-history-interface/blank2.html#test:2:3 document.write is not a function

  ▶ TIMEOUT [expected OK] /html/browsers/history/the-history-interface/002.html
  │ 
  │ Xlib:  extension "XFree86-VidModeExtension" missing on display ":0".
  └ ERROR:script::dom::bindings::error: Error at http://web-platform.test:8000/html/browsers/history/the-history-interface/blank2.html#test:2:3 document.write is not a function

  ▶ TIMEOUT [expected OK] /html/browsers/history/the-location-interface/location_reload.html
  │ 
  │ Xlib:  extension "XFree86-VidModeExtension" missing on display ":0".
  └ ERROR:script::dom::bindings::error: Error at http://web-platform.test:8000/html/browsers/history/the-location-interface/location_reload-iframe.html:1:3 parent._ping is not a function

  ▶ TIMEOUT [expected OK] /html/semantics/document-metadata/the-base-element/base_multiple.html
  │ 
  └ Xlib:  extension "XFree86-VidModeExtension" missing on display ":0".

  ▶ TIMEOUT [expected OK] /html/semantics/forms/form-submission-0/getactionurl.html
  │ 
  └ Xlib:  extension "XFree86-VidModeExtension" missing on display ":0".

  ▶ TIMEOUT [expected OK] /html/semantics/forms/form-submission-0/submit-entity-body.html
  │ 
  └ Xlib:  extension "XFree86-VidModeExtension" missing on display ":0".

  ▶ TIMEOUT [expected OK] /html/semantics/forms/form-submission-0/url-encoded.html
  │ 
  └ Xlib:  extension "XFree86-VidModeExtension" missing on display ":0".

  ▶ TIMEOUT [expected OK] /html/semantics/forms/the-button-element/button-activate.html
  │ 
  └ Xlib:  extension "XFree86-VidModeExtension" missing on display ":0".

  ▶ TIMEOUT [expected OK] /_mozilla/mozilla/FileAPI/file-upload.html
  │ 
  └ Xlib:  extension "XFree86-VidModeExtension" missing on display ":0".

  ▶ TIMEOUT [expected OK] /_mozilla/mozilla/form_submit_about.html
  │ 
  └ Xlib:  extension "XFree86-VidModeExtension" missing on display ":0".

  ▶ TIMEOUT [expected OK] /_mozilla/mozilla/htmllabel-activation.html
  │ 
  │ Xlib:  extension "XFree86-VidModeExtension" missing on display ":0".
  └ ERROR:script::dom::bindings::error: Uncaught exception: failed to extract information

  ▶ TIMEOUT [expected OK] /_mozilla/mozilla/mozbrowser/mozbrowser_click_fires_openwindow.html
  │ 
  └ Xlib:  extension "XFree86-VidModeExtension" missing on display ":0".

  ▶ TIMEOUT [expected OK] /_mozilla/mozilla/mozbrowser/mozbrowserlocationchange_event.html
  │ 
  └ Xlib:  extension "XFree86-VidModeExtension" missing on display ":0".
```

**Comment by aneeshusa** [comment]
@asajeffrey I'm having some trouble getting the last part (reducing `(PipelineId, PipelineId)` pairs to just one `PipelineId`), specifically (in `script_thread.rs`) how to get the browsing context associated with the parent pipeline id when we only have the chlid pipeline id. The approach currently in this PR - always using `self.root_browsing_context()` - doesn't work. I tried searching through the children browsing contexts to find one with a matching `PipelineId` and returning its parent, but it seems the child `PipelineId` won't necessarily have a matching browsing context. Is there another way to do this?

**Comment by aneeshusa** [comment]
@asajeffrey Ping on this, would love to get some help with the last part.

**Comment by aneeshusa** [comment]
Failures look like intermittents and linux-rel passed wpt ok. r? @asajeffrey

**Pull request merged**